src/plonk/mod.rs

```diff
@@ -485,39 +487,56 @@ impl<F: PrimeField> TableData<F> {
                 .chain(selector_columns)
                 .collect::<Vec<_>>()[..],
         );
-        let num_selector = self.cs.num_selectors();
-        let num_fixed = self.cs.num_fixed_columns();
-        let num_advice = self.cs.num_advice_columns();
 
-        // we use the same challenge y for combining custom gates and multiple lookup arguments
-        // need first commit all witness before generating y
-        let y = Expression::Polynomial(Query {
-            index: num_selector + num_fixed + num_advice,
-            rotation: Rotation(0),
-        });
         // suppose we have n polynomial expression: p_1,p_2,...,p_n
         // we combined them together as one: combined_poly = p_1*y^{n-1}+p_2*y^{n-2}+...+p_n
-        let combined_poly: MultiPolynomial<C::ScalarExt> = self
+        let num_gates = self
             .cs
             .gates()
             .iter()
             .flat_map(|gate| gate.polynomials().iter())
-            .map(|expr| Expression::from_halo2_expr(expr, num_selector, num_fixed))
-            .fold(Expression::Constant(F::ZERO), |acc, expr| {
-                Expression::Sum(
-                    Box::new(expr),
-                    Box::new(Expression::Product(Box::new(acc), Box::new(y.clone()))),
-                )
-            })
-            .expand();
+            .collect::<Vec<_>>()
+            .len();
+        // total number of challenges, this will be different after we including lookup
+        let num_challenges = if num_gates > 1 { 2 } else { 1 };
+        let gate: MultiPolynomial<F> = if num_gates > 1 {
+            let y = Expression::Challenge(0);
+            self.cs
+                .gates()
+                .iter()
+                .flat_map(|gate| gate.polynomials().iter())
+                .map(|expr| {
+                    Expression::from_halo2_expr(
+                        expr,
+                        self.cs.num_selectors(),
+                        self.cs.num_fixed_columns(),
+                    )
+                })
+                .fold(Expression::Constant(F::ZERO), |acc, expr| {
+                    Expression::Sum(
+                        Box::new(expr),
+                        Box::new(Expression::Product(Box::new(acc), Box::new(y.clone()))),
+                    )
+                })
```

Comment:
Do we really need to `sum` with zero at first iteration? We can use [`reduce` ](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.reduce). 

And thanks to this we can remove the branching above altogether, because the logic for one element is similar.

Reply:
Probably we can remove the zero. Can you explain the use of `reduce` here?

As a side note:
As you can see, I differentiate the case of one element or multiple elements. This is because if we don't differentiate as before, then the variable `y` will always be registered in the MultiPolynomial and we always include y in the challenges even it is never been used. This will cost 1 extra hash operation in IVC circuit, which is not optimal.